### PR TITLE
Introduce Role Filtering API to AuthServer CRD Identity Providers

### PR DESCRIPTION
**Pull request by ddubson**
Which other branches do you want a technical writer to cherry-pick this PR to (if any)?

It's best to PR to the most recent relevant branch and leave all the cherry-picking to the
writer, except where earlier branches require factual changes to the content.
For more information about the branches, see https://github.com/pivotal/docs-tap#branches

### CODE REVIEW

File: app-sso/getting-started/appsso-overview.hbs.md [hunk 1]
```diff
@@ -27,19 +27,26 @@ etc.). Here is an example of an `id_token` as issued by an Authorization Server:
...
+	"family_name": "Doe",
+	"email": "jane.doe@example.com",
+	"roles": [
+		"developer",
+		"org-user"
+	]
 }
 ```
 
+> **Note** `roles` claim is only part of an `id_token` when user roles are mapped. More info on mapping for
```
Comment (Kehrlann) [inline app-sso/getting-started/appsso-overview.hbs.md:47]: `roles` are only present if they exist *and* the `roles` scope was requested in the authorization request.

```suggestion
> **Note** `roles` claim can only be part of an `id_token` when user roles are mapped. More info on mapping for
```

File: app-sso/service-operators/identity-providers.hbs.md [hunk 2]
```diff
@@ -148,14 +189,51 @@ Where:
 - `user.seachFilter` is the filter for LDAP search. It must contain the string `{0}`, which is replaced
   by the `dn` of the user when performing a search. For example, when logging in with the username `marie`, the filter for LDAP
   search is `cn=marie`.
-- `group` (optional) defaults to unset. It configures how LDAP groups are mapped to user roles in the `id_token` claims.
+- `roles` (optional) defaults to unset. It configures how LDAP groups are mapped to user roles in the `id_token` claims.
+  If not set, the user has no roles.
+  - `fromUpstream.attribute` selects which attribute of the group entry are mapped to a user role. If an attribute has multiple
+    values, the first value is selected.
+  - `fromUpstream.search` (optional) toggles "Active Directory" search and uses recursive search to find groups for a given user.
+- `group` (DEPRECATED, optional) defaults to unset. It configures how LDAP groups are mapped to user roles in the `id_token` claims.
```
Comment (Kehrlann) [inline app-sso/service-operators/identity-providers.hbs.md:197]:
```suggestion
- `group` (DEPRECATED, use `role` instead, optional) defaults to unset. It configures how LDAP groups are mapped to user roles in the `id_token` claims.
```

File: app-sso/service-operators/identity-providers.hbs.md [hunk 3]
```diff
@@ -148,14 +189,51 @@ Where:
...
   If not set, the user has no roles.
-  - `group.roleAttriubte` selects which attribute of the group entry are mapped to a user role. If an attribute has multiple
+  - `group.roleAttribute` selects which attribute of the group entry are mapped to a user role. If an attribute has multiple
     values, the first value is selected.
   - `group.search` (optional) toggles "Active Directory" search and uses recursive search to find groups for a given user.
 
 Verify the configuration by visiting the `AuthServer`'s issuer URI in your browser and log in with the username and password from LDAP.
 
+### <a id='openid-external-groups-mapping'></a> External groups mapping
```
Comment (Kehrlann) [inline app-sso/service-operators/identity-providers.hbs.md:205]: Anchor issue
```suggestion
### <a id='ldap-external-groups-mapping'></a> External groups mapping
```

And this should probably be merged with `ActiveDirectory group search`, it's functionally the same example but with more explanations.

**Review by Kehrlann — COMMENTED** [review]
Nice progress.

Watch out, `Identity Providers` > `LDAP` >   all examples still use `spec.identityProviders.ldap.group`

File: app-sso/service-operators/identity-providers.hbs.md [hunk 4]
```diff
@@ -148,18 +199,65 @@ Where:
...
   search is `cn=marie`.
-- `group` (optional) defaults to unset. It configures how LDAP groups are mapped to user roles in the `id_token` claims.
+- `roles` (optional) defaults to unset. It configures how LDAP groups are mapped to user roles in the `id_token` claims.
   If not set, the user has no roles.
-  - `group.roleAttriubte` selects which attribute of the group entry are mapped to a user role. If an attribute has multiple
+    - `fromUpstream.attribute` selects which attribute of the group entry are mapped to a user role. If an attribute has
+      multiple
+      values, the first value is selected.
+    - `fromUpstream.search` (optional) toggles "Active Directory" search and uses recursive search to find groups for a
+      given user.
```
Comment (Kehrlann) [inline app-sso/service-operators/identity-providers.hbs.md:208]: **[issue]** this is openldap group search, not ActiveDirectory.

>  - `fromUpstream.search` (optional) toggles "OpenLDAP"-style group search, optionally uses recursive search to find groups for a given user.

File: app-sso/service-operators/identity-providers.hbs.md [hunk 5]
```diff
@@ -148,18 +199,65 @@ Where:
...
     values, the first value is selected.
   - `group.search` (optional) toggles "Active Directory" search and uses recursive search to find groups for a given user.
 
 Verify the configuration by visiting the `AuthServer`'s issuer URI in your browser and log in with the username and password from LDAP.
 
+### <a id='ldap-external-groups-mapping'></a> External groups mapping
+
+To configure retrieval of user roles as part of a claim within an identity token, you may map the LDAP identity
+provider's group attribute to AppSSO's `roles` non-standard OpenID Connect claim. An identity token issued by AppSSO
+will have `roles` claim attached when mapped.
```
Comment (Kehrlann) [inline app-sso/service-operators/identity-providers.hbs.md:223]: **[suggestion]** More LDAP-specific:

> To configure retrieval of user roles as part of a claim within an identity token, you may map the LDAP identity provider's "user groups" to AppSSO's `roles` non-standard OpenID Connect claim. An identity token issued by AppSSO will have `roles` claim attached when mapped.

File: app-sso/service-operators/identity-providers.hbs.md [hunk 6]
```diff
@@ -541,7 +677,129 @@ There are multiple options for generating bcrypt hashes:
...
+
+### <a id='roles-filters'></a> Filters
+
+Available filters are:
+
+- `exactMatch` - match the groups exactly. This filter is **case-sensitive**. e.g. `exactMatch: "developer"` will match
+  only the group named "developer" and no other.
+- `regex` - match groups according to the defined regular expression pattern. The regular expression provided should not
+  be surrounded by forward slashes (`/`), and should only contain the pattern (e.g. `.*`).
+  This filter is **case-insensitive**. e.g. `regex: ^admin` will match groups starting with the word "admin".
```
Comment (Kehrlann) [inline app-sso/service-operators/identity-providers.hbs.md:715]: **[question]** should we say this is actually a [Golang regular expression](https://pkg.go.dev/regexp)? 

> The regexp world is dark and full of terrors.

**Review by Kehrlann — COMMENTED** [review]
Massive work. Congratulations 👍 

A few small issues, and a suggestion around adding diagrams - maybe that's part of another PR?

File: app-sso/crds/authserver.hbs.md [hunk 7]
```diff
@@ -130,10 +144,24 @@ spec:
           name: ""
         scopes:
           - ""
+        claimMappings:
+          roles: "" # DEPRECATED, use 'openID.roles.fromUpstream.claim' instead.
```
[the remark this answers is not in the document]
Comment (ddubson) [reply app-sso/crds/authserver.hbs.md:148]: Oh interesting, I didn't quite understand that, for `openid`, `roles` was the only mapping feasible.

File: app-sso/service-operators/identity-providers.hbs.md [hunk 8]
```diff
@@ -69,14 +70,53 @@ You can also run `curl -s "https://openid.example.com/.well-known/openid-configu
...
+- `my-oidc-provider-groups` claim from the ID token issued by `my-oidc-provider` is mapped into the `roles` claim in id
+  tokens issued by AppSSO.
 
 Verify the configuration by visiting the `AuthServer`'s issuer URI in your browser and select `my-oidc-provider`.
 
+### <a id='openid-external-groups-mapping'></a> External groups mapping
+
+To configure retrieval of user roles as part of a claim within an identity token, you may map the OpenID identity
+provider's groups claim to AppSSO's `roles` non-standard OpenID Connect claim. An identity token issued by
+AppSSO will have `roles` claim attached when mapped.
```
[the remark this answers is not in the document]
Comment (ddubson) [reply app-sso/service-operators/identity-providers.hbs.md]: Added subsection for explanation around roles claim mapping and filtering

**Pull request closed without merging**